### ISSUE TITLE
a11y: Respect prefers-reduced-motion for all animations

### ISSUE THREAD

**Issue by t-hamano** [WordPress/gutenberg#68282]
## What problem does this address?

If we use transition or animation CSS, we are expected to use the `reduce-motion` mixin. Example:

```scss
.selector {
	transition: all 400ms linear;
	@include reduce-motion("transition");
}
```

This code compiles to the following to respect the user's preference and disable animations:

```css
.selector {
	transition: all 400ms linear;
}
@media (prefers-reduced-motion:reduce){
  .selector {
    transition-delay:0s;
    transition-duration:0s;
  }
}
```

However, there are many places where `transition` or `animation` is used without this mixin.

## What is your proposed solution?

We could update stylelint to enforce this rule, but since this is a custom rule, we would need to add our own plugin. I'm not sure if it would be worth creating a custom plugin to enforce this rule. For now, I think it's best to find the missing mixin and add it.

### Steps to get started

1. Search for `transition: ` or `animation: ` in the `.scss` files in the Gutenberg project.
2. Make sure the value is something other than `none`.
4. Check to see if the `reduce-motion` mixin exists. If it doesn't exist, we may need one.
6. Identify the component that has that CSS applied.
8. In your Chrome browser, change the "Emulate CSS media feature prefers-reduced-motion" setting to `reduce`.
  https://developer.chrome.com/docs/devtools/rendering/emulate-css#emulate_css_media_feature_prefers-reduced-motion
9. Do something to that component that will cause a transition or animation (such as hover or focus).
10. If transition or animation occurs, it means the user's animation preference is not being respected, so add the mixin.

---

Update: We aim to use standard patterns across projects rather than using mixins.

```css
@media not ( prefers-reduced-motion ) {
	transition: all 400ms linear;
}
```

### Submitted PRs

- [x] base-styles: https://github.com/WordPress/gutenberg/pull/68427
- [x] block-directory: https://github.com/WordPress/gutenberg/pull/68420
- [x] block-editor: https://github.com/WordPress/gutenberg/pull/68417
- [x] block-library: https://github.com/WordPress/gutenberg/pull/68315
- [x] components: https://github.com/WordPress/gutenberg/pull/68421
- [x] customize-widgets: https://github.com/WordPress/gutenberg/pull/68425
- [x] data-views: https://github.com/WordPress/gutenberg/pull/68422
- [x] edit-post: https://github.com/WordPress/gutenberg/pull/68426
- [x] edit-site: https://github.com/WordPress/gutenberg/pull/68419
- [x] edit-widgets: https://github.com/WordPress/gutenberg/pull/68427
- [x] editor: https://github.com/WordPress/gutenberg/pull/68424
- [x] nux: https://github.com/WordPress/gutenberg/pull/68423

**Comment by yogeshbhutkar** [comment]
I’d like to work on this issue and prepare a corresponding PR 🚀.

**Comment by himanshupathak95** [comment]
Hey @t-hamano, Thanks for raising this.

I will raise a PR to fix this.

**Comment by t-hamano** [comment]
@himanshupathak95 Thanks for addressing this issue.

If you don't mind, could you provide some screenshots of which components each of the reduce-motion mixins affects so we can make sure they work as expected?

Also, you don't need to tackle them all at once. If there are too many to tackle, you can tackle them package by package.

**Comment by himanshupathak95** [comment]
@t-hamano, I am thinking of doing this package by package since there are a lot of places that need to change and there must be distinction amongst them to test and discuss each package correctly in isolation.

I am thinking maybe we can create separate PRs to track packages specifically and execute the changes so we don't break anything. Please let me know what you think.

**Comment by t-hamano** [comment]
Yes, I think it's a good idea to submit a PR for each package 👍

**Comment by mirka** [comment]
No strong opinion, but I wanted to note that in `@wordpress/components` we decided to adopt this standard pattern instead of a mixin:

```css
@media not ( prefers-reduced-motion ) {
	transition: all 400ms linear;
}
```

It isn't any less ergonomic, it's immediately understandable without knowledge of a custom mixin implementation, and we end up with fewer lines of CSS. Would it be worth considering this pattern for the wider codebase as well?

**Comment by t-hamano** [comment]
@mirka Thanks for the feedback!

> Would it be worth considering this pattern for the wider codebase as well?

That's a good idea. I think it might be better to proceed in the following order:

1. If the `reduce-motion` mixin itself is missing, wrap the transition or animation in the media query.
2. Replace where the `reduce-motion` mixin is already used with the media query.

**Comment by t-hamano** [comment]
> ```css
> @media not ( prefers-reduced-motion ) {
> 	transition: all 400ms linear;
> }
> ```

It turns out that this formatting isn't available in the sass version used by Gutenberg.

I've submitted a PR to update the sass: https://github.com/WordPress/gutenberg/pull/68380

**Comment by Mayank-Tripathi32** [comment]
@t-hamano I will work on other blocks in the block-library, one at a time. This way, each change will have a clear pull request, making testing and finding regressions easier.

Tracking List:

[ x ] - Block Library: `block-library`- #68413
[ x ] - Block Editor: `block-editor` - #68417 - needs testing

**Comment by t-hamano** [comment]
@Mayank-Tripathi32 Can the block library PRs be consolidated into one or two PRs? Otherwise, there will be too many PRs. For other packages, if the changes are small, I think it's ok to consolidate multiple packages into one PR.

**Comment by himanshupathak95** [comment]
@t-hamano, In [my former PR](https://github.com/WordPress/gutenberg/pull/68315), as I mentioned earlier, I have consolidated all the changes to the block library. I think it's better to consolidate the changes else as you correctly mentioned there will be too many PRs with small changes.

**Comment by Mayank-Tripathi32** [comment]
> [@Mayank-Tripathi32](https://github.com/Mayank-Tripathi32) Can the block library PRs be consolidated into one or two PRs? Otherwise, there will be too many PRs. For other packages, if the changes are small, I think it's ok to consolidate multiple packages into one PR.
 
@t-hamano  cherry-picked all the branches into #68413, please have a look

**Comment by t-hamano** [comment]
⚠️ Note: An issue was reported where the Storybook build would fail when updating to `@media not (prefers-reduced-motion)`.

https://github.com/WordPress/gutenberg/pull/68419#issuecomment-2567550864

If you work on this issue, I recommend that you also check that `storybook:dev` works properly at the same time to avoid the same issue. Also, #68466 has been submitted to check Storybook build on CI.

**Comment by t-hamano** [comment]
Thanks to everyone working on this issue.

I've scanned the whole project, and once all the current dubmitted PRs are merged, all the work should be done.

@Mayank-Tripathi32 @im3dabasia Feel free to let us know when you're ready to publish your PR.

**Comment by Mayank-Tripathi32** [comment]
> Thanks to everyone working on this issue.
> 
> I've scanned the whole project, and once all the current dubmitted PRs are merged, all the work should be done.
> 
> [@Mayank-Tripathi32](https://github.com/Mayank-Tripathi32) [@im3dabasia](https://github.com/im3dabasia) Feel free to let us know when you're ready to publish your PR.

@t-hamano I encountered some difficulty in locating all the component implementations for testing. Do you have any suggestions or tools to help test the PRs more efficiently?

**Comment by t-hamano** [comment]
@Mayank-Tripathi32 If the component that the CSS is applied to is exposed in Storybook, you'll be able to use that to test it, if not, you'll probably need to parse the code to figure out where it's actually used.

**Comment by t-hamano** [comment]
Thanks to everyone working on this issue.

All tasks are completed, but I found one last piece of code that needs refactoring:

https://github.com/WordPress/gutenberg/blob/8b837f7e2dd6e144f689017fafeaac472cead637/packages/block-library/src/social-links/editor.scss#L89-L93

Does anyone want to work on this?

**Comment by himanshupathak95** [comment]
Yes @t-hamano, I would like to take that up.
